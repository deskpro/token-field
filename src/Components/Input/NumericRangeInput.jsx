import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import { Input } from '@deskpro/react-components';
import styles from '../../styles/style.css';
import TokenInput from './TokenInput';

export default class NumericRangeInput extends TokenInput {
  static moveCaretAtEnd(e) {
    const tempValue = e.target.value;
    e.target.value = '';
    e.target.value = tempValue;
  }

  constructor(props) {
    super(props);
    let from = '';
    let to = '';
    if (this.props.token.value && this.props.token.value.length) {
      from = this.props.token.value.range[0];
    }
    if (this.props.token.value && this.props.token.value.length > 1) {
      to = this.props.token.value.range[1];
    }
    this.state = {
      from,
      to,
    };
  }

  renderInput = () => {
    const { unitPhrase } = this.props;
    const { from, to } = this.state;
    const fromValue = (from) ? this.props.convertFromValue(from) : '';
    const toValue = (to) ? this.props.convertFromValue(to) : '';
    return (
      <span>
        <Input
          value={fromValue}
          name="from"
          className={classNames(styles.input)}
          onChange={this.handleChange}
          onKeyDown={this.handleKeyDown}
          ref={(c) => { this.inputFrom = c; }}
          onFocus={NumericRangeInput.moveCaretAtEnd}
          placeholder="__"
        />
        {unitPhrase} to
        <Input
          value={toValue}
          name="to"
          className={classNames(styles.input)}
          onChange={this.handleChange}
          onKeyDown={this.handleKeyDown}
          ref={(c) => { this.inputTo = c; }}
          onFocus={NumericRangeInput.moveCaretAtEnd}
          placeholder="__"
        />
        {unitPhrase}
      </span>
    );
  };

  renderValue = () => {
    const { unitPhrase } = this.props;
    const { from, to } = this.state;
    let displayFrom;
    let displayTo;
    if (from) {
      displayFrom = this.props.convertFromValue(from);
    } else {
      displayFrom = '__';
    }
    if (to) {
      displayTo = this.props.convertFromValue(to);
    } else {
      displayTo = '__';
    }
    return `${displayFrom} ${unitPhrase} to ${displayTo} ${unitPhrase}`;
  };

  focus(end) {
    super.focus(end);
  }

  focusInput = (end) => {
    this.props.onFocus(this.props.tokenKey);
    if (end) {
      this.inputTo.focus();
    } else {
      this.inputFrom.focus();
    }
  };

  handleChange = (value, name) => {
    this.setState({
      [name]: this.props.convertToValue(value)
    });
  };

  onBlur = () => {
    this.props.onBlur();
    this.props.onChange({ range: [this.state.from, this.state.to] });
  };

  handleKeyDown = (e) => {
    switch (e.key) {
      case 'Escape':
        this.setState({
          value: this.props.token.value
        });
        this.disableEditMode();
        break;
      case 'Tab':
      case 'Enter':
        if (!e.shiftKey) {
          if (e.target.name === 'from') {
            this.inputTo.focus();
          } else {
            this.props.selectNextToken();
            this.disableEditMode();
          }
        } else if (e.target.name === 'to') {
          this.inputFrom.focus();
        } else {
          this.props.selectPreviousToken();
          this.disableEditMode();
        }
        break;
      default:
        // Do nothing
        return;
    }
    e.preventDefault();
  };
}


NumericRangeInput.propTypes = {
  ...TokenInput.propTypes,
  token: PropTypes.shape({
    type:  PropTypes.string,
    value: PropTypes.array,
  }).isRequired,
  unitPhrase:       PropTypes.string,
  convertToValue:   PropTypes.func,
  convertFromValue: PropTypes.func,
};
NumericRangeInput.defaultProps = {
  ...TokenInput.defaultProps,
  unitPhrase: '',
  convertToValue(v) { return v; },
  convertFromValue(v) { return v; },
};
