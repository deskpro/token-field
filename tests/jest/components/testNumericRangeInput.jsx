import React from 'react';
import renderer from 'react-test-renderer';
import NumericRangeInput from 'Components/Input/NumericRangeInput';
import { noop } from '@deskpro/react-components/dist/utils';

const convertFromValue = value => Math.round(value / 1024 / 1024);
const convertToValue = value => value * 1024 * 1024;

it('+++capturing Snapshot of NumericRangeInput', () => {
  const token = {
    type:  'user-message',
    value: {
      range: [12582912, 20971520]
    },
  };
  const renderedValue = renderer.create(
    <NumericRangeInput
      token={token}
      unitPhrase="MB"
      className="test"
      selectPreviousToken={noop}
      selectNextToken={noop}
      removeToken={noop}
    />
  ).toJSON();
  expect(renderedValue).toMatchSnapshot();
});
it('+++capturing Snapshot of NumericRangeInput with custom convert', () => {
  const token = {
    type:  'user-message',
    value: {
      range: [12582912, 20971520]
    },
  };
  const renderedValue = renderer.create(
    <NumericRangeInput
      token={token}
      unitPhrase="MB"
      className="test"
      convertToValue={convertToValue}
      convertFromValue={convertFromValue}
      selectPreviousToken={noop}
      selectNextToken={noop}
      removeToken={noop}
    />
  ).toJSON();
  expect(renderedValue).toMatchSnapshot();
});
it('+++capturing Snapshot of NumericRangeInput with empty value', () => {
  const token = {
    type:  'user-message',
    value: []
  };
  const renderedValue = renderer.create(
    <NumericRangeInput
      token={token}
      unitPhrase="MB"
      className="test"
      selectPreviousToken={noop}
      selectNextToken={noop}
      removeToken={noop}
    />
  ).toJSON();
  expect(renderedValue).toMatchSnapshot();
});
