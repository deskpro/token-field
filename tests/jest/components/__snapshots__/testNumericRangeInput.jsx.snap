// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`+++capturing Snapshot of NumericRangeInput 1`] = `
<div
  className="token test"
>
  <div>
    <div
      className="dp-code label"
    >
      :
    </div>
    <span
      className="value"
      onClick={[Function]}
    >
      __ MB to __ MB
    </span>
  </div>
  <div
    className="token-remove remove"
    onClick={[Function]}
  >
    X
  </div>
</div>
`;

exports[`+++capturing Snapshot of NumericRangeInput with custom convert 1`] = `
<div
  className="token test"
>
  <div>
    <div
      className="dp-code label"
    >
      :
    </div>
    <span
      className="value"
      onClick={[Function]}
    >
      __ MB to __ MB
    </span>
  </div>
  <div
    className="token-remove remove"
    onClick={[Function]}
  >
    X
  </div>
</div>
`;

exports[`+++capturing Snapshot of NumericRangeInput with empty value 1`] = `
<div
  className="token test"
>
  <div>
    <div
      className="dp-code label"
    >
      :
    </div>
    <span
      className="value"
      onClick={[Function]}
    >
      __ MB to __ MB
    </span>
  </div>
  <div
    className="token-remove remove"
    onClick={[Function]}
  >
    X
  </div>
</div>
`;
